File: Difficulty: Hard/Burning Tree/burning-tree.cpp
/*
class Node {
  public:
    int data;
    Node *left;
    Node *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};
*/
class Solution {
  public:
    void markparents(Node* root,unordered_map<Node*,Node*> &parent_track,Node* target_node){
        queue<Node*> q;
        q.push(root);
        while(!q.empty()){
            Node* curr = q.front();
            q.pop();
            if(curr->left){
                parent_track[curr->left] = curr;
                q.push(curr->left);
            }
            if(curr->right){
                parent_track[curr->right] = curr;
                q.push(curr->right);
            }
        }
    }
  
  
    Node* istarget(Node* root,int target){
        if (root==NULL){
            return NULL;
        }
        if(root->data==target){
            return root;
        }
        Node* left_res = istarget(root->left,target);
        if(left_res!=NULL) return left_res;
        
        Node* right_res = istarget(root->right,target);
        return right_res;
    
    }
    
    int minTime(Node* root, int target) {
        // code here
        Node* target_node = istarget(root,target);
        unordered_map<Node*,Node*> parent_track;
        markparents(root,parent_track,target_node);
        
        queue<Node*> q;
        unordered_map<Node*,int> vis;
        q.push(target_node);
        vis[target_node]= 1;
        int maxt = 0;
        while(!q.empty()){
            int size = q.size();
            int fl = 0;
            
            for(int i=0;i<size;i++){
                auto curr = q.front();
                q.pop();
                if (curr->left && !vis[curr->left]){
                    fl =1;
                    vis[curr->left]= 1;
                    q.push(curr->left);
                }
                if (curr->right && !vis[curr->right]){
                    fl =1;
                    vis[curr->right]= 1;
                    q.push(curr->right);
                }
                if(parent_track[curr] && !vis[parent_track[curr]]){
                    fl=1;
                    vis[parent_track[curr]] = 1;
                    q.push(parent_track[curr]);
                }
            }
            if (fl) maxt++;
        }
        return maxt;
    }
};

File: Difficulty: Hard/Burning Tree/burning-tree.py
from collections import deque, defaultdict
class Solution:
    
    def minTime(self, root, target):
        # code here 
        if root is None:
            return -1
        
        q= deque([root])
        mpp={root:None}
        tar=0
        while q:
            curr= q.popleft()
            if curr.data== target:
                tar=curr
            
            if curr.left:
                mpp[curr.left]=curr
                q.append(curr.left)
            if curr.right:
                mpp[curr.right]=curr
                q.append(curr.right)
        
        vis={}
        ans=-1
        q.append(tar)
        while q:
            size = len(q)
            for i in range(size):
                curr= q.popleft()
                vis[curr]=True
                
                if curr.left and not vis.get(curr.left, False):
                    q.append(curr.left)
                if curr.right and not vis.get(curr.right,False):
                    q.append(curr.right)
                    
                parent = mpp[curr]
                if parent and not vis.get(parent,False):
                    q.append(parent)
                
            ans+=1
        return ans
            
